lrtable/src/lib/statetable.rs

```diff
@@ -306,10 +305,45 @@ where
         })
     }
 
+    fn decode(bits: u32) -> Action<StorageT> { // XXX howto find StorageT? encode in int as well?
+        //           Action   State/Prod
+        let mask = 0b00000000_11111111;
```

Comment:
An 8 bit mask for 4 values is quite wasteful: we can compress this down to 2 bits. I would expect something like this:

```
let kind = bits & 0b11;
let val = (bits & ~0b11) >> 2;
```

should work OK?

Reply:
Indeed, my plan was to start simple at and get it to work first and then get some  comments from you before optimising this. I was just wondering what type we should use. Do you think `u16` is enough or still too much?

Reply:
It's not trivial, because we need to represent `StorageT::max_value() << 2`: if someone uses `u16` for `StorageT`, we'll be running a risk of crashing. I would probably use a `u64` (with an appropriate `assert`... just in case!), because we'll later use an integer packing vec, at which point it doesn't really matter what type we're storing.

Reply:
Isn't `let val = (bits & ~0b11) >> 2;` the same as `let val = bits >> 2;`? Also, I think your version assumes that the action type is encoded at the end of the integer, however I designed it to be at the beginning. Maybe that was a bad choice? Should I swap them around?

Reply:
Oops, you're quite right. Yes `bits >> 2` is fine. Putting the type bits starting at position 0 is definitely the right way to go, since that leaves a variable number of bits for the "data".

Reply:
I understand now! That makes indeed more sense to have the unknown sized number at the beginning and the fixed size at the end!